File: Sorting_algorithms/python/src/quick_sort.py
def partition(arr: int, lowest_index: int, highest_index: int) -> int:
    pivot:int = arr[highest_index]
    left_index:int = lowest_index

    for i in range(lowest_index, highest_index):
        if arr[i] <= pivot:
            arr[left_index], arr[i] = arr[i], arr[left_index]
            left_index+=1
    arr[left_index], arr[highest_index] = arr[highest_index], arr[left_index]
    return left_index

def quick_sort(arr: int, lowest_index: int, highest_index: int) -> list:
    if lowest_index < highest_index:
        pivot_index:int = partition(arr=arr, lowest_index=lowest_index, highest_index=highest_index)
        quick_sort(arr, lowest_index, pivot_index-1)
        quick_sort(arr, pivot_index+1, highest_index)
    return arr

if __name__ == "__main__":
    arr = [11, 30, -1, 2, 7, 12, 13, 20, 8]
    print(f"\nBefore Sort: {arr}")
    print(f"After Sort: {quick_sort(arr=arr, lowest_index=0, highest_index=len(arr)-1)}\n")
   

File: Sorting_algorithms/Kotlin/src/quick_sort.kt
fun swap_values(arr:IntArray, index_01: Int, index_02: Int) {
    var temp: Int = arr[index_01]
    arr[index_01] = arr[index_02]
    arr[index_02] = temp
}

fun partition (arr: IntArray, lowest_index: Int, highest_index: Int): Int {
    val pivot: Int = arr[highest_index]
    var left_index: Int = lowest_index
    for (i in lowest_index until highest_index) {
        if (arr[i] <= pivot) {
            swap_values(arr, left_index, i)
            left_index++
        }
    }
    swap_values(arr, left_index, highest_index)
    return left_index
}

fun quick_sort(arr: IntArray, lowest_index: Int, highest_index: Int): IntArray {
    if (lowest_index < highest_index) {
        val pivot_index = partition(arr, lowest_index, highest_index)
        quick_sort(arr, lowest_index, pivot_index-1)
        quick_sort(arr, pivot_index+1, highest_index)
    }
    return arr
}


fun main () {
    var arrList = intArrayOf(11, 30, -1, 2, 7, 12, 13, 20, 8)
    println("\nBefore Sort: ${arrList.joinToString(", ")}")
    arrList = quick_sort(arrList, 0, arrList.size-1)
    println("After Sort: ${arrList.joinToString(", ")}")
}